### ISSUE TITLE
/etc/hostname is incorrect when hostname is set via cloud-config

### ISSUE THREAD

**Issue by Moeser** [rancher/os#867]
I set our hostname via cloud-config with a file like this:

```
#cloud-config
hostname: frontend1
```

Running the 'hostname' binary shows the correct hostname.  However, when I cat /etc/hostname it still shows the hostname as rancher.

```
[rancher@frontend1 ~]$ cat /etc/hostname
rancher
```

This is breaking functionality such as DDNS via DHCP.  All of our rancher hosts get a DHCP lease from the DHCP server as "rancher" instead of their actual hostname.  Note that in RancherOS the ddns hostname option for dhcpcd.conf is configured by default, but the dhcpcd.conf documentation shows that they read the value from /etc/hostname when the hostname option is set.

Might also be related to other tickets like #829 though their problem description is slightly different than what we're seeing.

**Comment by janeczku** [comment]
What's the data source for your cloud-config?
Does this problem persist even after rebooting the host?
Or after you reload the network config via `sudo ros service restart network`?

> but the dhcpcd.conf documentation shows that they read the value from /etc/hostname when the hostname option is set.

The value in /etc/hostname is irrelevant here. DHCPD gets the hostname via kernel syscall and not by parsing the file.

**Comment by joshwget** [comment]
Do you remember having these problems prior to v0.4.3? Even though it sounds a little different, I think the underlying issue is still #829.

**Comment by janeczku** [comment]
@joshwget I think these are different issues.
#829: hostname is not set from cloud-init
Here: hostname is correctly set from cloud-init (`rancher@frontend1`)

**Comment by joshwget** [comment]
Ah, okay. Thanks for the clarification @janeczku!

**Comment by Moeser** [comment]
> What's the data source for your cloud-config?

I'm not totally sure what you mean here.  I'm  passing it as a URL to the kernel on bootup.  I've checked /var/lib/rancher/conf/cloud-config.d/boot.yml on the booted system and it looks correct.

> Does this problem persist even after rebooting the host?

Yes

> Or after you reload the network config via sudo ros service restart network?

Yes.  The file doesn't change even after running sudo ros service restart network.  It still contains 'rancher'

> The value in /etc/hostname is irrelevant here. DHCPD gets the hostname via kernel syscall and not by parsing the file.

I wish this statement were correct, since as you pointed out, the system hostname (shown on the command prompt above) is correct.  Unfortunately, all of our rancher hosts are registering themselves as 'rancher' to the dhcp server.  Note that if I hand-edit /etc/hostname and run `sudo ros service restart network` then dhcpcd sends the correct hostname to the dhcp server and dns gets updated.  dhcpcd is clearly retrieving the value from /etc/hostname, not from the system uname() call.  

The documentation I'm referring to is here: https://wiki.archlinux.org/index.php/dhcpcd#DHCP_Client_Identifier  where it says `hostname sends the hostname set in /etc/hostname`

**Comment by janeczku** [comment]
> Yes. The file doesn't change even after running sudo ros service restart network. It still contains 'rancher'

Sorry if i wasn't clear. I thought the issue at hand is that your hosts identify themselves as "rancher" to the DHCP server, right? My question therefore was if this behavior changes once you reload the network or restart the host. I am pretty positive that the host will identify itself as "frontend-1" to the DHCP server at this point.

The fact that your hosts identify themselves as "rancher" on first boot is expected behavior: The cloud-config hostname is not set until a network connection is established (how else would it fetch the cloud-config from the URL you passed in the kernel cmd line?).

**Comment by Moeser** [comment]
> The question was if your hosts still send "rancher" as DHCP client identifier once you reload the network or restart the host.

I just tested this out, and they send the correct hostname to DHCPD after running `ros service restart network`, even when /etc/hostname is still set to 'rancher'.

**Comment by Moeser** [comment]
So since that works, as a temporary workaround, should I add a runcmd: to cloud-config to reload the network during init?  Or is that still before the hostname is set?  I'm not sure if the items in cloud-config arae processed sequentially or not.

**Comment by Moeser** [comment]
I just tried adding ros service network restart to the cloud-config but that didn't work.  Still looking for workarounds here.

**Comment by Moeser** [comment]
I tried adding a write_files: section to cloud-config which would write 'frontend1' to /etc/hostname, but that didn't work.  Somehow it's getting overwritten with 'rancher' every boot.  I even updated /etc/hostname manually to be absolutely sure it was correct before rebooting, and it still booted back up containing 'rancher'.

I had tried to use runcmd: to run ros service network restart during bootup, but that didn't do anything.  Trying out other commands makes me think that rancher just doesn't support runcmd yet, but I'm not sure.  I haven't been able to find a complete reference on what cloud-config directives rancher supports.

Anyhow, the original issue is still here.  IE: /etc/hostname isn't being updated to match the hostname: value in cloud-config.  In addition, even if it were updated correctly, it's becoming apparent that a network restart would have to happen after the hostname is set anyway, in order for the correct hostname to be sent to the dhcp server (regardless of whether /etc/hostname is updated).  So I guess there's actually two bugs here?

In addition to DDNS, sending the correct hostname to dhcp is also important for organizations that assign leases based on hostname.  They'll probably get the wrong IP if their lease is requested under the wrong hostname.  That's unavoidable for the first request, as you pointed out, but at least by the time the host is fully booted, it would have the correct IP.

**Comment by janeczku** [comment]
> I tried adding a write_files: section to cloud-config which would write 'frontend1' to /etc/hostname

Please stop worrying about `/etc/hostname`. As i *already* said, this file is absolutely irrelevant for anything discussed here. The hostname is set at boot to the value of the hostname option in rancher config (`sudo ros config export --full`).

> I just tried adding ros service network restart to the cloud-config but that didn't work. Still looking for workarounds here.

See here for a working `write_files` section that restarts the network on the first boot and will fix your DHCP problem: https://gist.github.com/janeczku/a44f533d9af80e620db7

**Comment by joshwget** [comment]
I think I agree with @janeczku here. `/etc/hostname` isn't a standard, which means applications really shouldn't be reading from it. I don't see much value in keeping it in sync with the true hostname, but I'm not really against it either.

Another alternative to using `write_files` is to upgrade to v0.4.4, where networking is reapplied by default.

**Comment by Moeser** [comment]
You're probably right that we can drop this issue if /etc/hostname isn't being used for anything in RancherOS.

Just one quick thought though: support /could/ be added for /etc/hostname to preserve the hostname across reboots.  That would fix a problem we're seeing where a host always identifies itself as 'rancher' on the first dhcp request, then after cloud-config re-requests a lease identifies itself as the real hostname.  From our server side, the fact that the host is always 'rancher' during boot is causing a mess of DNS updates every time a host reboots (removing the real hostname and PTR and adding the 'rancher' hostname and PTR inititally on boot, then removing the 'rancher' hostname and adding the real hostname and PTR back again after cloud-config).  Just food for thought.

I'll go ahead and resolve this, since /etc/hostname doesn't appear to be used.

**Comment by joshwget** [comment]
Issues with hostname like the one you mentioned should be fixed in v0.4.5. Setting hostname in the cloud config in RancherOS should behave like setting hostname in `/etc/hostname` in other distros. If it's not, then it's really a bug with cloud config and should be fixed there rather than introducing another configuration file. Thanks for following up!

**Comment by segator** [comment]
I'm using rancherOS 0.5.0 and I have the same problem, it was fixed?

**Comment by deniseschannon** [comment]
@segator Yes, it's fixed in v0.5.0. I just re-verified. 

Can you share your cloud-config that you used?

I used this

```
#cloud-config
ssh_authorized_keys:
  - mysshkey
hostname: testing
```

After logging in, when I did `cat /etc/hostname` it came up as `testing`.
